### PR TITLE
working on perturbation detectors

### PR DESCRIPTION
**Pull request by rabah-khalek**
- For now only coloring and blurring are supported for models that have "predict_gray_image" implemented (only hf models for now)
- To add later: resizing by half but this requires also rescaling the labels of the object_detection

We can keep the scope limited to coloring and blurring for now. @bmalezieux maybe you can support `predict_gray_image` for the rest of the models we have?

### CODE REVIEW

File: giskard_vision/core/detectors/perturbation.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+
+import cv2
+
+from giskard_vision.core.dataloaders.wrappers import FilteredDataLoader
+from giskard_vision.core.detectors.base import (
+    DetectorVisionBase,
+    IssueGroup,
+    ScanResult,
+)
+from giskard_vision.landmark_detection.tests.base import TestDiff
```
Comment (bmalezieux) [inline giskard_vision/core/detectors/perturbation.py:14]: I think we should move TestDiff in core instead of importing it from landmark_detection

**Review by bmalezieux — CHANGES_REQUESTED** [review]
We can also add a NoisePerturbationDetector to evaluate robustness to noise

File: giskard_vision/core/detectors/perturbation.py [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+            os.makedirs(f"{current_path}/examples_images", exist_ok=True)
+            filename_examples = []
+
+            index_worst = 0 if test_result.indexes_examples is None else test_result.indexes_examples[0]
+
+            if isinstance(dl, FilteredDataLoader):
+                filename_example_dataloader_ref = str(Path() / "examples_images" / f"{dataset.name}_{index_worst}.png")
+                cv2.imwrite(
+                    filename_example_dataloader_ref, cv2.resize(dataset[index_worst][0][0], (0, 0), fx=0.3, fy=0.3)
+                )
```
[the remark this answers is not in the document]
Comment (rabah-khalek) [reply giskard_vision/core/detectors/perturbation.py:80]: ideally a bunch of values in the range of [0.1, 0.8] would be nice yes

File: giskard_vision/core/detectors/perturbation.py [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+
+        return results
+
+    def get_scan_result(self, test_result, filename_examples, name, size_data) -> ScanResult:
+        try:
+            from giskard.scanner.issues import IssueLevel
+        except (ImportError, ModuleNotFoundError) as e:
+            raise GiskardImportError(["giskard"]) from e
+
+        relative_delta = (test_result.metric_value_test - test_result.metric_value_ref) / test_result.metric_value_ref
```
[the remark this answers is not in the document]
Comment (rabah-khalek) [reply giskard_vision/core/detectors/perturbation.py:96]: agreed

File: giskard_vision/core/detectors/perturbation.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (bmalezieux) [reply giskard_vision/core/detectors/perturbation.py:68]: Actually this is only to save the transformed images in order to display them in the report

File: giskard_vision/core/detectors/perturbation.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (rabah-khalek) [reply giskard_vision/core/detectors/perturbation.py:84]: taken care of in https://github.com/Giskard-AI/giskard-vision/pull/53

**Pull request merged**